source/common/upstream/load_balancer_impl.cc

```diff
@@ -1299,25 +1299,89 @@ HostConstSharedPtr LeastRequestLoadBalancer::unweightedHostPick(const HostVector
                                                                 const HostsSource&) {
   HostSharedPtr candidate_host = nullptr;
 
+  switch (selection_method_) {
+    case envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest_SelectionMethod_FULL_SCAN:
+      candidate_host = unweightedHostPickFullScan(hosts_to_use);
+      break;
+    case envoy::extensions::load_balancing_policies::least_request::v3::LeastRequest_SelectionMethod_N_CHOICES:
+    default:
+      candidate_host = unweightedHostPickNChoices(hosts_to_use);
+  }
+
+  return candidate_host;
+}
+
+HostSharedPtr LeastRequestLoadBalancer::unweightedHostPickFullScan(const HostVector& hosts_to_use) {
+  HostSharedPtr candidate_host = nullptr;
+
+  // "Randomly" scan the host array using a linear congruential generator with the following parameters:
+  // - The "modulus" is the number of hosts
+  // - The "multiplier" is 1
+  // - The "increment" is a non-zero value less than the modulus.
+  //   It is guaranteed non-zero because it is determined by:
+  //     random large prime number / modulus ... which never be 0
+  // - The "start value" is a random index in the host array
+  //
+  // With these parameters, the generator has a period equal to the modulus (number of hosts).
+  //
+  // The effect of this algorithm is to visit each index of the host array exactly once.
+  // The next index will always be +"increment" indices from the current index.
+  //
+  // This algorithm is not *very* random, but it is random *enough* for its purpose.
+  // Approaches with more randomness were not used because they are less performant for larger host arrays
+  // (e.g., requiring the allocation of an array the length of the host array).
+
+  unsigned long random_value = random_.random();
+
+  // Choose a random large prime greater than the host array length to ensure the "increment"
+  // can be any value between 0 and the array length.
+  // By using a random large prime in each scan, the "increment" can differ between scans,
+  // and therefore the sequences of indices visited can differ between scans. This behavior is
+  // important to avoid selecting the exact same host every time if multiple hosts tie for
+  // least requests.
+  unsigned long large_prime = LARGE_PRIME_CHOICES[random_value % LARGE_PRIME_CHOICES_LEN];
+
+  // Because the initial value of host_scan_index will always be the last index scanned,
+  // it must be a random index.
+  unsigned long host_scan_index = random_value % hosts_to_use.size();
```

Comment:
Just in case, I should probably used different `random_value`s here and in the large_prime choice.